feat: adding new metric: up down sum observer (#1272)
open-telemetry/opentelemetry-js: packages/opentelemetry-metrics/README.md

### Value Observer

Choose this kind of metric when only last value is important without worry about aggregation.
The callback can be sync or async.

```js
const { MeterProvider } = require('@opentelemetry/metrics');

const meter = new MeterProvider().getMeter('your-meter-name');

// async callback - for operation that needs to wait for value
meter.createValueObserver('your_metric_name', {

>   description: 'Example of a async observer with callback',

Reviewer: ```suggestion
  description: 'Example of an async observer with callback',
```

nit

}, async (observerResult) => {
  const value = getAsyncValue();

>   observerResult.observe(value, { label: '1' });

Reviewer 2: Looks like this should be `observerResult.observe(await value, { label: '1' });`

Reviewer 3: Think moving the await to `const value = await getAsyncValue();` is more idiomatic

});

function getAsyncValue() {
  return new Promise((resolve) => {
    setTimeout(()=> {
      resolve(Math.random());
    }, 100);
  });
}

> // sync callback in case you don't need to wait for value

Reviewer: 👍 
Great to provide both examples.

meter.createValueObserver('your_metric_name', {
  description: 'Example of a async observer with callback',
}, (observerResult) => {
  observerResult.observe(getRandomValue(), { label: '1' });
  observerResult.observe(getRandomValue(), { label: '2' });
});

function getRandomValue() {
  return Math.random();
}
```

### UpDownSumObserver

> Choose this kind of metric when sum is important.

Reviewer 2: Adding something like `The observe method may accept either positive or negative values.` or something to similar effect might be helpful.

The callback can be sync or async.

```js
const { MeterProvider } = require('@opentelemetry/metrics');

const meter = new MeterProvider().getMeter('your-meter-name');

> // async callback - for operation that needs to wait for value

Reviewer: ```suggestion
// async callback - for operations that need to wait for value
```

nit: I find this reads slightly better imho

Author: one operation will wait for one `value`, if there were more then it would be `for operations that need to wait for values`

meter.createUpDownSumObserver('your_metric_name', {
  description: 'Example of an async observer with callback',
}, async (observerResult) => {
  const value = getAsyncValue();

>   observerResult.observe(value, { label: '1' });

Reviewer 2: Likewise with hitting `value` with an `await`

Reviewer 3: Would rather see it on the above line.

Author: yea I missed that :)

### Value Observer

Choose this kind of metric when only last value is important without worry about aggregation.

> Callback can return either nothing or promise to support async operation

Reviewer 4: ```suggestion
The callback can return either nothing or promise to support async operation.
```

```js
const { MeterProvider } = require('@opentelemetry/metrics');

const meter = new MeterProvider().getMeter('your-meter-name');

// callback with promise - for operations that needs to wait for value
meter.createValueObserver('cpu_core_usage', {
  description: 'Example of a async observer with callback',
}, (observerResult) => {
  return new Promise((resolve) => {
    getAsyncValue().then((value) => {
      observerResult.observe(value, { core: '1' });
      resolve();
    });
  });
});

function getAsyncValue() {
  return new Promise((resolve) => {
    setTimeout(()=> {
      resolve(Math.random());
    }, 100);
  });
}

// callback without promise in case you don't need to wait for value
meter.createValueObserver('cpu_core_usage', {
  description: 'Example of a async observer with callback',
}, (observerResult) => {
  observerResult.observe(getRandomValue(), { core: '1' });
  observerResult.observe(getRandomValue(), { core: '2' });
});

function getRandomValue() {
  return Math.random();
}
```

> ### Up Down Sum Observer

Reviewer 4: ```suggestion
### UpDownSumObserver
```

Reviewer 4: One word.

Choose this kind of metric when sum is important.
The callback can return either nothing or promise to support async operation.

```js
const { MeterProvider } = require('@opentelemetry/metrics');

const meter = new MeterProvider().getMeter('your-meter-name');

// callback with promise - for operations that needs to wait for value

> meter.createUpDownSumObserver('cpu_core_usage', {

Reviewer 3: It doesn't really make sense to add cpu core usage does it?

  description: 'Example of an async observer with callback',

> }, (observerResult) => {
>   return new Promise((resolve) => {
>     getAsyncValue().then((value) => {
>       observerResult.observe(value, { core: '1' });
>       resolve();
>     });
>   });

Reviewer 3: Might be less confusing to just make the callback async

```suggestion
}, async (observerResult) => {
  const value = await getAsyncValue();
  observerResult.observe(value, { core: '1' });
```

});

function getAsyncValue() {
  return new Promise((resolve) => {
    setTimeout(()=> {
      resolve(Math.random());
    }, 100);
  });
}

// callback without promise in case you don't need to wait for value
meter.createUpDownSumObserver('cpu_core_usage', {
  description: 'Example of an async observer with callback',
}, (observerResult) => {
  observerResult.observe(getRandomValue(), { core: '1' });

>   resolve();

Reviewer 3: resolve should be removed
